File: bindings/java/examples/zebrad-regtest/src/main/java/com/zcash/t2z/examples/ZebraClient.java
/**
 * Zebra RPC Client
 *
 * Pure Java HttpClient-based JSON-RPC client for Zebra - minimal dependencies.
 */
package com.zcash.t2z.examples;

import com.google.gson.Gson;
import com.google.gson.JsonArray;
import com.google.gson.JsonElement;
import com.google.gson.JsonObject;

import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.util.Base64;
import java.util.concurrent.TimeUnit;

public class ZebraClient {
    private final String url;
    private final String authHeader;
    private int idCounter = 0;
    private final Gson gson = new Gson();
    private final HttpClient httpClient = HttpClient.newHttpClient();

    public ZebraClient() {
        this(null, null);
    }

    public ZebraClient(String host, Integer port) {
        String h = host != null ? host : System.getenv("ZEBRA_HOST");
        if (h == null) h = "localhost";

        int p = port != null ? port : 18232;
        String portEnv = System.getenv("ZEBRA_PORT");
        if (portEnv != null && port == null) {
            try {
                p = Integer.parseInt(portEnv);
            } catch (NumberFormatException ignored) {
            }
        }

        this.url = "http://" + h + ":" + p;

        String user = System.getenv("RPC_USER");
        String pass = System.getenv("RPC_PASSWORD");
        if (user != null && pass != null) {
            this.authHeader = "Basic " + Base64.getEncoder().encodeToString((user + ":" + pass).getBytes());
        } else {
            this.authHeader = null;
        }
    }

    /**
     * Make a raw JSON-RPC call.
     */
    private JsonElement rawCall(String method, JsonArray params) throws Exception {
        JsonObject request = new JsonObject();
        request.addProperty("jsonrpc", "2.0");
        request.addProperty("method", method);
        request.add("params", params != null ? params : new JsonArray());
        request.addProperty("id", ++idCounter);

        HttpRequest.Builder requestBuilder = HttpRequest.newBuilder()
                .uri(URI.create(url))
                .header("Content-Type", "application/json")
                .POST(HttpRequest.BodyPublishers.ofString(gson.toJson(request)));

        if (authHeader != null) {
            requestBuilder.header("Authorization", authHeader);
        }

        HttpResponse<String> response = httpClient.send(requestBuilder.build(),
                HttpResponse.BodyHandlers.ofString());

        if (response.statusCode() < 200 || response.statusCode() >= 300) {
            throw new Exception("HTTP error: " + response.statusCode());
        }

        JsonObject jsonResponse = gson.fromJson(response.body(), JsonObject.class);

        JsonElement error = jsonResponse.get("error");
        if (error != null && !error.isJsonNull()) {
            throw new Exception("RPC error: " + error);
        }

        return jsonResponse.get("result");
    }

    /**
     * Wait for the node to be ready.
     */
    public void waitForReady(int maxAttempts, long delayMs) throws Exception {
        for (int i = 0; i < maxAttempts; i++) {
            try {
                getBlockchainInfo();
                return;
            } catch (Exception e) {
                if (i == maxAttempts - 1) {
                    throw new Exception("Node not ready after " + maxAttempts + " attempts: " + e.getMessage());
                }
                Thread.sleep(delayMs);
            }
        }
    }

    /**
     * Get blockchain info.
     */
    public BlockchainInfo getBlockchainInfo() throws Exception {
        JsonObject result = rawCall("getblockchaininfo", null).getAsJsonObject();
        return new BlockchainInfo(
                result.get("chain").getAsString(),
                result.get("blocks").getAsInt(),
                result.get("bestblockhash").getAsString()
        );
    }

    /**
     * Get current block count.
     */
    public int getBlockCount() throws Exception {
        return rawCall("getblockcount", null).getAsInt();
    }

    /**
     * Get block hash by height.
     */
    public String getBlockHash(int height) throws Exception {
        JsonArray params = new JsonArray();
        params.add(height);
        return rawCall("getblockhash", params).getAsString();
    }

    /**
     * Get block by hash.
     */
    public JsonObject getBlock(String hash, int verbosity) throws Exception {
        JsonArray params = new JsonArray();
        params.add(hash);
        params.add(verbosity);
        return rawCall("getblock", params).getAsJsonObject();
    }

    /**
     * Send a raw transaction.
     */
    public String sendRawTransaction(String hexString) throws Exception {
        JsonArray params = new JsonArray();
        params.add(hexString);
        return rawCall("sendrawtransaction", params).getAsString();
    }

    /**
     * Get raw transaction.
     */
    public JsonElement getRawTransaction(String txid, boolean verbose) throws Exception {
        JsonArray params = new JsonArray();
        params.add(txid);
        params.add(verbose ? 1 : 0);
        return rawCall("getrawtransaction", params);
    }

    /**
     * Wait for blocks to reach target height.
     */
    public int waitForBlocks(int targetHeight, long timeoutMs) throws Exception {
        long startTime = System.currentTimeMillis();
        int lastHeight = 0;

        while (System.currentTimeMillis() - startTime < timeoutMs) {
            BlockchainInfo info = getBlockchainInfo();
            if (info.blocks >= targetHeight) {
                return info.blocks;
            }
            if (info.blocks != lastHeight) {
                lastHeight = info.blocks;
                System.out.println("  Block height: " + info.blocks);
            }
            Thread.sleep(1000);
        }

        throw new Exception("Timeout waiting for height " + targetHeight);
    }

    public void close() {
        // HttpClient doesn't need explicit close
    }

    /**
     * Blockchain info data class.
     */
    public static class BlockchainInfo {
        public final String chain;
        public final int blocks;
        public final String bestblockhash;

        public BlockchainInfo(String chain, int blocks, String bestblockhash) {
            this.chain = chain;
            this.blocks = blocks;
            this.bestblockhash = bestblockhash;
        }
    }
}

File: kotlin/examples/zebrad-t2z/src/main/kotlin/com/zcash/t2z/examples/ZebraClient.kt
/**
 * Zebra RPC Client
 *
 * Pure Java HttpClient-based JSON-RPC client for Zebra - minimal dependencies.
 */
package com.zcash.t2z.examples

import kotlinx.coroutines.delay
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.*
import java.net.URI
import java.net.http.HttpClient
import java.net.http.HttpRequest
import java.net.http.HttpResponse
import java.util.Base64

@Serializable
data class BlockchainInfo(
    val chain: String,
    val blocks: Int,
    val headers: Int? = null,
    val bestblockhash: String,
    val difficulty: Double? = null,
    val verificationprogress: Double? = null
)

@Serializable
data class UTXO(
    val txid: String,
    val vout: Int,
    val address: String,
    val scriptPubKey: String,
    val amount: Long, // in zatoshis
    val height: Int
)

class ZebraClient(
    host: String? = null,
    port: Int? = null
) {
    private val url: String
    private val authHeader: String?
    private var idCounter = 0
    private val json = Json { ignoreUnknownKeys = true }
    private val httpClient = HttpClient.newHttpClient()

    init {
        val h = host ?: System.getenv("ZEBRA_HOST") ?: "localhost"
        val p = port ?: System.getenv("ZEBRA_PORT")?.toIntOrNull() ?: 18232
        url = "http://$h:$p"

        val user = System.getenv("RPC_USER")
        val pass = System.getenv("RPC_PASSWORD")
        authHeader = if (user != null && pass != null) {
            "Basic " + Base64.getEncoder().encodeToString("$user:$pass".toByteArray())
        } else null
    }

    /**
     * Make a raw JSON-RPC call
     */
    private inline fun <reified T> rawCall(method: String, params: JsonArray = buildJsonArray {}): T {
        val requestBody = buildJsonObject {
            put("jsonrpc", "2.0")
            put("method", method)
            put("params", params)
            put("id", ++idCounter)
        }.toString()

        val requestBuilder = HttpRequest.newBuilder()
            .uri(URI.create(url))
            .header("Content-Type", "application/json")
            .POST(HttpRequest.BodyPublishers.ofString(requestBody))

        authHeader?.let { requestBuilder.header("Authorization", it) }

        val response = httpClient.send(requestBuilder.build(), HttpResponse.BodyHandlers.ofString())

        if (response.statusCode() !in 200..299) {
            throw Exception("HTTP error: ${response.statusCode()}")
        }

        val jsonResponse = json.parseToJsonElement(response.body()).jsonObject

        val error = jsonResponse["error"]
        if (error != null && error != JsonNull) {
            throw Exception("RPC error: $error")
        }

        return json.decodeFromJsonElement(jsonResponse["result"]!!)
    }

    /**
     * Wait for the node to be ready
     */
    suspend fun waitForReady(maxAttempts: Int = 30, delayMs: Long = 1000) {
        repeat(maxAttempts) { i ->
            try {
                getBlockchainInfo()
                return
            } catch (e: Exception) {
                if (i == maxAttempts - 1) {
                    throw Exception("Node not ready after $maxAttempts attempts: ${e.message}")
                }
                delay(delayMs)
            }
        }
    }

    /**
     * Get blockchain info
     */
    fun getBlockchainInfo(): BlockchainInfo = rawCall("getblockchaininfo")

    /**
     * Get current block count
     */
    fun getBlockCount(): Int = rawCall("getblockcount")

    /**
     * Get block hash by height
     */
    fun getBlockHash(height: Int): String = rawCall("getblockhash", buildJsonArray { add(height) })

    /**
     * Get block by hash
     */
    fun getBlock(hash: String, verbosity: Int = 1): JsonObject =
        rawCall("getblock", buildJsonArray { add(hash); add(verbosity) })

    /**
     * Send a raw transaction
     */
    fun sendRawTransaction(hexString: String): String =
        rawCall("sendrawtransaction", buildJsonArray { add(hexString) })

    /**
     * Get raw transaction
     */
    fun getRawTransaction(txid: String, verbose: Boolean = false): JsonElement =
        rawCall("getrawtransaction", buildJsonArray { add(txid); add(if (verbose) 1 else 0) })

    /**
     * Wait for blocks to reach target height
     */
    suspend fun waitForBlocks(targetHeight: Int, timeoutMs: Long = 120000): Int {
        val startTime = System.currentTimeMillis()
        var lastHeight = 0

        while (System.currentTimeMillis() - startTime < timeoutMs) {
            val info = getBlockchainInfo()
            if (info.blocks >= targetHeight) {
                return info.blocks
            }
            if (info.blocks != lastHeight) {
                lastHeight = info.blocks
                println("  Block height: ${info.blocks}")
            }
            delay(1000)
        }

        throw Exception("Timeout waiting for height $targetHeight")
    }

    fun close() {
        // HttpClient doesn't need explicit close
    }
}
